native-authentication/src/main/java/com/okta/nativeauthentication/NativeAuthenticationClient.kt

```diff
@@ -24,58 +24,76 @@ import com.okta.nativeauthentication.form.FormFactory
 import com.okta.nativeauthentication.form.FormTransformer
 import com.okta.nativeauthentication.form.RetryFormBuilder
 import com.okta.nativeauthentication.form.transformer.SignInTitleTransformer
-import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.SendChannel
 import kotlinx.coroutines.channels.awaitClose
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.channelFlow
 
-class NativeAuthenticationClient private constructor() {
+class NativeAuthenticationClient internal constructor(
+    private val formTransformers: List<FormTransformer>,
+    private val idxResponseTransformer: IdxResponseTransformer,
+) {
     companion object {
         init {
             SdkVersionsRegistry.register(SDK_VERSION)
         }
 
-        private var formTransformers: List<FormTransformer> = listOf(SignInTitleTransformer())
-
         fun create(
             callback: Callback,
-            interactionCodeFlowFactory: suspend () -> OidcClientResult<InteractionCodeFlow>,
+            interactionCodeFlowFactory: suspend () -> OidcClientResult<InteractionCodeFlow>
         ): Flow<Form> {
-            suspend fun ProducerScope<Form>.start(formFactory: FormFactory) {
-                when (val result = interactionCodeFlowFactory()) {
-                    is OidcClientResult.Error -> {
-                        formFactory.emit(
-                            RetryFormBuilder.create(this) {
-                                start(formFactory)
-                            }
-                        )
-                    }
-                    is OidcClientResult.Success -> {
-                        val interactionCodeFlow = result.result
-                        val wrapperCallback = object : Callback() {
-                            override fun signInComplete(token: Token) {
-                                super.signInComplete(token)
-                                callback.signInComplete(token)
-                                close()
-                            }
+            return NativeAuthenticationClient(
+                formTransformers = listOf(SignInTitleTransformer()),
+                idxResponseTransformer = RealIdxResponseTransformer(),
+            ).create(callback, interactionCodeFlowFactory)
+        }
+    }
+
+    abstract class Callback {
+        open fun signInComplete(token: Token) {
+        }
+    }
```

Comment:
Didn't notice this earlier, but why is this an abstract class instead of an interface?

Reply:
I want to be able to add methods to this in the future. I _could_ do this as an interface with default methods. But this seems to make my intent more explicit.